### PR TITLE
Configurable force pushing

### PR DESCRIPTION
**Pull request by jracollins**
Thought I would have a crack at my own issue I raised, so this is a work in progress PR to implement #205 

I'm not very familiar with Go and the test cases are a little fiddly to write (given most of the "actual" work is done by libgit2 and a single character change to the refspec) so will take me a little longer to write those up. I thought I'd open the draft PR so someone else doesn't unnecessarily pick up the issue when I'm almost done. 

In the meantime a couple questions: 
-  It isn't a breaking API change, but should I be implementing for the older api specs (v1alpha1) etc like I have done? (Not sure what the correct process is) 
-  Any issues with/comments on the actual implementation below?

### CODE REVIEW

**Comment by stefanprodan** [comment]
> should I be implementing for the older api specs 

No, older API versions are considered readonly. 

PS. Please signoff your commits, see https://github.com/fluxcd/image-automation-controller/pull/206/checks?check_run_id=3208778477

File: api/v1beta1/git.go [hunk 1]
```diff
@@ -87,4 +87,6 @@ type PushSpec struct {
 	// starting point, if it doesn't already exist.
 	// +required
 	Branch string `json:"branch"`
+	// Force specifies wether to force push the branch to origin
+	Force bool `json:"force"`
```
Comment (stefanprodan) [inline api/v1beta1/git.go:91]: This should be:

```go
	// This flag tells the controller to force push the branch to origin. Defaults to false.
	// +optional
	Force bool `json:"suspend,omitempty"`
```

Comment (jracollins) [reply api/v1beta1/git.go]: I'm assuming you mean just the `omitempty` part and keeping the key as `force`, not `suspend` - just pushed fix

Comment (stefanprodan) [reply api/v1beta1/git.go]: Yes thank you for changing it.

**Comment by squaremo** [comment]
This is going to be pretty tricky to test. The controller always does a fresh clone and checkout of the push branch; to necessitate a `push --force`, you would need to alter the upstream push branch _after_ the controller has fetched it, but before it tries to push to it. The test will fail, because it doesn't do this interleaving, but expects `--force` to be necessary.

```go
// observe the first commit
					waitForNewHead(localRepo, pushBranch)

					// Create commit on push branch that would conflict with a commit on top of the a branch from the checkout branch
					dummyCommitMessage := "Dummy Commit"
					commitInRepo(cloneLocalRepoURL, pushBranch, dummyCommitMessage, func(tmp string) {})

// origin pushBranch now has a commit "Dummy Commit"

					waitForNewHead(localRepo, pushBranch)

					head, err := localRepo.Reference(plumbing.NewRemoteReferenceName(originRemote, pushBranch), true)
					Expect(err).NotTo(HaveOccurred())
					pushBranchInitialHeadHash := head.String()
					dummyCommit, err := localRepo.CommitObject(head.Hash())
					Expect(err).NotTo(HaveOccurred())
					Expect(dummyCommit.Message).To(Equal(dummyCommitMessage))

// the local clone now (verifiably) has the head at Dummy Commit

					// update the policy and expect another commit in the push branch, deleting local branch to ensure commit is not on top of the dummy commit
					localRepo.DeleteBranch(pushBranch)
					policy.Status.LatestImage = "helloworld:v1.3.0"
					Expect(k8sClient.Status().Update(context.TODO(), policy)).To(Succeed())

// This deletes pushBranch locally, but does nothing to the origin

// When the policy object is updated, the controller will clone the repo from the origin, check out the branch specified, and commit on top of it.

					// Delete local push branch and refetch remote
					localRepo.DeleteBranch(pushBranch)
					waitForNewHead(localRepo, pushBranch)
					head, err = localRepo.Reference(plumbing.NewRemoteReferenceName(originRemote, pushBranch), true)
					Expect(err).NotTo(HaveOccurred())
					fetchedPushBranchLatestHeadHash := head.String()
					pushBranchLatestCommit, err := localRepo.CommitObject(head.Hash())
					Expect(err).NotTo(HaveOccurred())

// this will get the new head of push branch, which is the commit made by the controller (on top of Dummy Commit)

					// Expect push to fail
					// Check with messages for easier debugging
					Expect(dummyCommit.Message).To(Equal((pushBranchLatestCommit.Message)))

// this fails because the head commit is the commit made by the controller

					// Check with hashes for certainty
					Expect(pushBranchInitialHeadHash).To(Equal((fetchedPushBranchLatestHeadHash)))
```

**Pull request closed without merging**